src/neo/Consensus/ConsensusContext.cs

```diff
@@ -386,9 +386,21 @@ public void Reset(byte viewNumber)
                 CommitPayloads = new ConsensusPayload[Validators.Length];
                 if (LastSeenMessage == null)
                 {
-                    LastSeenMessage = new int[Validators.Length];
+                    LastSeenMessage = new Dictionary<ECPoint, int>();
                     for (int i = 0; i < Validators.Length; i++)
-                        LastSeenMessage[i] = -1;
+                        LastSeenMessage[Validators[i]] = (int)Snapshot.Height;
+                }
+                if (0 < Snapshot.Height && Blockchain.Singleton.GetBlock(Snapshot.Height - 1).NextConsensus != Blockchain.Singleton.GetBlock(Snapshot.Height).NextConsensus)
+                {
+                    var previous_last_seen_message = LastSeenMessage;
```

Comment:
Maybe this part could be a function in order to leave ConsensusContext more clear.

Reply:
done.